sway/desktop/output.c

```diff
@@ -754,9 +754,87 @@ static void render_container(struct sway_output *output,
 	case L_TABBED:
 		render_container_tabbed(output, damage, con, parent_focused);
 		break;
-	case L_FLOATING:
-		// TODO
-		break;
+	}
+}
+
+static bool floater_intersects_output(struct sway_container *floater,
+		struct sway_container *output) {
+	struct wlr_box box = {
+		.x = floater->x,
+		.y = floater->y,
+		.width = floater->width,
+		.height = floater->height,
+	};
+	return wlr_output_layout_intersects(root_container.sway_root->output_layout,
+			output->sway_output->wlr_output, &box);
+}
+
+static void container_translate(struct sway_container *con, int x, int y) {
+	con->x += x;
+	con->y += y;
+	if (con->type == C_VIEW) {
+		con->sway_view->x += x;
+		con->sway_view->y += y;
+	} else {
+		for (int i = 0; i < con->children->length; ++i) {
+			struct sway_container *child = con->children->items[i];
+			container_translate(child, x, y);
+		}
+	}
+}
+
+static void render_floating_container(struct sway_output *soutput,
+		pixman_region32_t *damage, struct sway_container *con) {
+	// We need to translate the floating container's coordinates from layout
+	// coordinates into output-local coordinates. This needs to happen for all
+	// children of the floating container too.
+	struct sway_container *output = container_parent(con, C_OUTPUT);
+	container_translate(con, -output->x, -output->y);
+
+	if (con->type == C_VIEW) {
+		struct sway_view *view = con->sway_view;
+		struct sway_seat *seat = input_manager_current_seat(input_manager);
+		struct sway_container *focus = seat_get_focus(seat);
+		struct border_colors *colors;
+		struct wlr_texture *title_texture;
+		struct wlr_texture *marks_texture;
+
+		if (focus == con) {
+			colors = &config->border_colors.focused;
+			title_texture = con->title_focused;
+			marks_texture = view->marks_focused;
+		} else {
+			colors = &config->border_colors.unfocused;
+			title_texture = con->title_unfocused;
+			marks_texture = view->marks_unfocused;
+		}
+		render_titlebar(soutput, damage, con, con->x, con->y, con->width,
```

Comment:
There should be a check against the border style here. If you have a floating view with opacity set below 1.0 and `border pixel` or `border none`, the title bar is visible since it is being rendered underneath the view.